### ISSUE TITLE
Research about C-RNN-GAN's Implementation

### ISSUE THREAD

**Issue by halil-erdogan** [hedonistrh/TurkishMusicGeneration#38]
**One-line summary (optional)**
- Try their source code, if does not work, implement it. 🏭 

**Reason**
- We would like to reproduce it.

**Design**
- Firstly try their source code directly with our midi
- If it does not work as our expected, try to understand its architecture
- Provide code snippet about architecture

**Acceptance Criteria**
- We should have consensus about how we move with C-RNN-GAN.

**Outcome**
- Experiment result of their implementation and probably summary-code snippet about their method.

**Comment by halil-erdogan** [comment]
- [Official Implementation](https://github.com/olofmogren/c-rnn-gan)
- C-RNN-GAN based [_Pokemon Music Generation_](https://towardsdatascience.com/generating-pokemon-inspired-music-from-neural-networks-bc240014132?gi=4a4ed82686b9)

**Comment by halil-erdogan** [comment]
- I will dive into [Source Code](https://github.com/olofmogren/c-rnn-gan/blob/master/rnn_gan.py#L222).

- This is the discriminator and generator code from _Pokemon Music Generation_, however, some hyper-parameters can be not same with the source code of official implementation. I will check it asap. 🖲 

```python
def build_discriminator(self):
    model = Sequential()
    model.add(LSTM(512, input_shape = self.seq_shape,
                return_sequences=True))
    model.add(Bidirectional(LSTM(512)))
    model.add(Dense(512))
    model.add(LeakyReLU(alpha=0.2))
    model.add(Dense(256))
    model.add(LeakyReLU(alpha=0.2))
    model.add(Dense(1, activation='sigmoid'))
    model.summary()

    seq = Input(shape=self.seq_shape)
    validity = model(seq)
    
    return Model(seq, validity)

def build_generator(self):
    model = Sequential()
    model.add(Dense(256, input_dim=self.latent_dim))
    model.add(BatchNormalization(momentum=0.8))
    model.add(Dense(512))
    model.add(LeakyReLU(alpha=0.2))
    model.add(BatchNormalization(momentum=0.8))
    model.add(Dense(1024))
    model.add(LeakyReLU(alpha=0.2))
    model.add(BatchNormalization(momentum=0.8))
    model.add(Dense(np.prod(self.seq_shape), 
                activation='tanh'))
    model.add(Reshape(self.seq_shape))
    model.summary()

    noise = Input(shape=(self.latent_dim,))
    seq = model(noise)

    return Model(noise, seq)
```

**Comment by halil-erdogan** [comment]
This [post](https://www.wouterbulten.nl/blog/tech/getting-started-with-generative-adversarial-networks/) will be helpful.

I think the implementation of C-RNN-GAN is more or less same with the following codes. _(Not the exact one. We are using different decay and loss function.)_
```python
def build_discriminator():
    model = Sequential()
    model.add(Bidirectional(LSTM(350, activation="relu", 
                                return_sequences=True, 
                                kernel_initializer='random_normal',
                                bias_initializer='random_normal'),
                                kernel_regularizer=regularizers.l2(1.0),
                                input_shape=(seq_length, vocab_size)))
    model.add(Dropout(0.5))
    model.add(Dense((dense_size, activation="relu"))
    model.add(Dropout(0.5))
    model.add(Dense(1, activation='sigmoid'))
    model.summary()
    return model


def build_generator():
    model = Sequential()
    model.add(LSTM(350, input_shape = (seq_length, vocab_size),
                activation="relu",
                kernel_initializer='glorot_uniform',
                bias_initializer=initializers.Constant(0),
                kernel_regularizer=regularizers.l2(1.0),
                return_sequences=True))    
    model.add(Dropout(0.3))
    model.add(Dense(350))
    model.add(Activation('softmax'))
    return model

net_discriminator = discriminator()
net_generator = generator()


# Optimizer of Discriminator
optim_discriminator = SGD(lr=0.1, clipvalue=5.0, decay=0.001)
model_discriminator = Sequential()
model_discriminator.add(net_discriminator)
model_discriminator.compile(loss='binary_crossentropy', optimizer=optim_discriminator, metrics=['accuracy'])

# Optimizer for Generator
optim_adversarial = Adam(lr=0.1, clipvalue=5.0, decay=0.001)
model_adversarial = Sequential()
model_adversarial.add(net_generator)

# Disable layers in discriminator
for layer in net_discriminator.layers:
    layer.trainable = False

model_adversarial.add(net_discriminator)
model_adversarial.compile(loss='binary_crossentropy', optimizer=optim_adversarial, metrics=['accuracy'])
```
